// @bfmeta/node-sdk/build/src/api/atom_api/transactions/atom_transaction/commonTransactionApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonTransactionApi = void 0;
const _transactionApi_1 = require("./_transactionApi");
require("../../../../constants");
class CommonTransactionApi extends _transactionApi_1.TransactionApi {
    GENERATE_API_PATH = "/generate";
    CREATE_API_PATH = "";
    PACKAGE_API_PATH = "";
    BROADCAST_API_PATH = "";
    async broadcastCompleteTransaction(argv) {
        const apiPath = this.getApiPath();
        try {
            const result = await this.networkHelper.sendPostRequest(apiPath, argv);
            return result;
        }
        catch (e) {
            const errorInfo = {
                success: false,
                error: {
                    code: "7001",
                    message: `request api ${apiPath} error`,
                    description: e.message,
                },
            };
            return errorInfo;
        }
    }
}
exports.CommonTransactionApi = CommonTransactionApi;
