// @bnqkl/util-node/cjs/platform/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.platformInfo = exports.isWebWorker = exports.isWebMainThread = exports.isIOS = exports.isAndroid = exports.isDenoRuntime = exports.isWebView = exports.isCordova = void 0;
const global_js_1 = require("../typings/global.js");
const diff_node_js_1 = require("./diff@node.js");
const checkType = (name, type) => {
    try {
        return new Function(`return typeof ${name} === "${type}"`)();
    }
    catch (_) {
        return false;
    }
};
exports.isCordova = checkType("cordova", "object");
/**web worker and main thread all has location as navigator */
exports.isWebView = checkType("navigator", "object");
exports.isDenoRuntime = checkType("Deno", "object");
exports.isAndroid = exports.isWebView && /Android/i.test(navigator.userAgent);
exports.isIOS = exports.isWebView && /iPhone|iPod|iPad/i.test(navigator.userAgent);
exports.isWebMainThread = exports.isWebView && checkType("document", "object");
exports.isWebWorker = exports.isWebView && !exports.isWebMainThread;
exports.platformInfo = {
    getGlobalFlag(flag_name, defaultValue = "") {
        const g = exports.isDenoRuntime ? this : this.global();
        return (g[flag_name] ||
            (g.process && g.process.env && g.process.env[flag_name]) ||
            (g.location &&
                g.location.href &&
                new URL(g.location.href).searchParams.get(flag_name)) ||
            (g.localStorage && g.localStorage.getItem(flag_name)) ||
            defaultValue);
    },
    global() {
        return global_js_1.$global;
    },
    platformName() {
        if (diff_node_js_1.isNodejs) {
            return "Nodejs";
        }
        const device_name = exports.isAndroid ? "Android" : exports.isIOS ? "IOS" : "unknown";
        if (exports.isCordova) {
            return "Cordova-" + device_name;
        }
        if (exports.isWebMainThread) {
            return "WebMaster-" + device_name;
        }
        if (exports.isWebWorker) {
            return "WebWorker-" + device_name;
        }
        return "UNKNOWN";
    },
    getChannel() {
        return "UNKNOWN";
    },
    getBusiness() {
        return "UNKNOWN";
    },
};
