// @bfmeta/node-sdk/build/test/transactions/beExchangeAnyMultiAll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "boost scorpion peanut output undo useful trash burden custom party click offer leisure magnet obscure drop gather blind predict walk since strike thumb minimum";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "20000",
            applyBlockHeight: 50,
            remark: { message: "create beExchangeAnyMulti" },
            transactionSignature: "b0517406378d6900208a278b70f0c0159bc814512771d3bf1db76a09a7181f63987ad38b290378f5f064e2be19f92527f269e32516c28b2907888e6edf9eee07",
            toExchangeAssets: [
                {
                    toExchangeSource: helpers_1.CHAIN_INFO.chainMagic,
                    toExchangeChainName: helpers_1.CHAIN_INFO.chainName,
                    toExchangeParentAssetType: 3,
                    toExchangeAssetType: "m_skyrim_dragonborn00",
                    toExchangeAssetPrealnum: "1",
                    taxInformation: {
                        taxCollector: "cLrUCNAWPyPH96bqqC3JQXZ3CtsvvXmNj1",
                        taxAssetPrealnum: "0",
                    },
                },
                {
                    toExchangeSource: helpers_1.CHAIN_INFO.chainMagic,
                    toExchangeChainName: helpers_1.CHAIN_INFO.chainName,
                    toExchangeParentAssetType: 5,
                    toExchangeAssetType: helpers_1.CHAIN_INFO.chainAssetType,
                    toExchangeAssetPrealnum: "100000",
                },
            ],
            beExchangeAssets: [
                {
                    beExchangeSource: helpers_1.CHAIN_INFO.chainMagic,
                    beExchangeChainName: helpers_1.CHAIN_INFO.chainName,
                    beExchangeParentAssetType: 5,
                    beExchangeAssetType: helpers_1.ASSET_INFO.assetType,
                    beExchangeAssetPrealnum: "100000",
                },
            ],
            recipientId: "cLrUCNAWPyPH96bqqC3JQXZ3CtsvvXmNj1",
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createBeExchangeAnyMultiAll(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastBeExchangeAnyMulti({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
