// @bfmeta/node-sdk/build/test/transactions/destroyAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "22222";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "2000",
            applyBlockHeight: 156500,
            remark: { message: "create destroyAsset" },
            assetType: "FTC",
            amount: "9720000000000000",
            recipientId: "c4PvHHmzpYCBEkaJ6hL8mJ5gVdMHwybJyu"
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createDestroyAsset(argv);
        console.log(createResult);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastDestroyAsset({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
