// @bfmeta/node-sdk/build/test/transactions/giftAny.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "upgrade jump sugar congress glare expect other firm morning donate motor pride minute frame amount chimney wood gallery twelve barely dose blame convince enhance";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "44000000",
            applyBlockHeight: 30,
            remark: { message: "create giftAny" },
            parentAssetType: 3,
            assetType: "biohazard_dragonborn",
            amount: "1",
            totalGrabableTimes: 1,
            taxInformation: {
                taxCollector: "cLrUCNAWPyPH96bqqC3JQXZ3CtsvvXmNj1",
                taxAssetPrealnum: "0",
            },
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createGiftAny(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastGiftAny({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
