// @bfmeta/node-sdk/build/test/transactions/setLnsRecordValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "upgrade jump sugar congress glare expect other firm morning donate motor pride minute frame amount chimney wood gallery twelve barely dose blame convince enhance";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "10000",
            applyBlockHeight: 50,
            remark: { message: "create setLnsRecordValue" },
            name: `hylq.${helpers_1.CHAIN_INFO.chainName}`,
            addRecord: [1..toString(), "192.168.111.119"],
            operationType: 0,
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createSetLnsRecordValue(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastSetLnsRecordValue({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
