// @bfmeta/node-sdk/build/test/transactions/toExchangeSpecialAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "upgrade jump sugar congress glare expect other firm morning donate motor pride minute frame amount chimney wood gallery twelve barely dose blame convince enhance";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "1000",
            applyBlockHeight: 50,
            remark: { message: "create toExchangeSpecialAsset" },
            toExchangeSource: helpers_1.CHAIN_INFO.chainMagic,
            toExchangeChainName: helpers_1.CHAIN_INFO.chainName,
            toExchangeAsset: `hylq.${helpers_1.CHAIN_INFO.chainName}`,
            beExchangeSource: helpers_1.CHAIN_INFO.chainMagic,
            beExchangeChainName: helpers_1.CHAIN_INFO.chainName,
            beExchangeAsset: "BFT",
            exchangeNumber: "66666",
            exchangeAssetType: 1,
            exchangeDirection: 0,
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createToExchangeSpecialAsset(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastToExchangeSpecialAsset({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
