// @bfmeta/node-sdk/build/test/transactions/toExchangeAnyMulti.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "upgrade jump sugar congress glare expect other firm morning donate motor pride minute frame amount chimney wood gallery twelve barely dose blame convince enhance";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "20000",
            applyBlockHeight: 40,
            remark: { message: "create toExchangeAnyMulti" },
            toExchangeAssets: [
                {
                    toExchangeParentAssetType: 3,
                    toExchangeAssetType: "m_skyrim_dragonborn00",
                    toExchangeAssetPrealnum: "1",
                    taxInformation: {
                        taxCollector: "cLrUCNAWPyPH96bqqC3JQXZ3CtsvvXmNj1",
                        taxAssetPrealnum: "0",
                    },
                },
                {
                    toExchangeParentAssetType: 5,
                    toExchangeAssetType: helpers_1.CHAIN_INFO.chainAssetType,
                    toExchangeAssetPrealnum: "100000",
                    assetExchangeWeightRatio: {
                        toExchangeAssetWeight: "1000",
                        beExchangeAssetWeight: "1",
                    },
                },
            ],
            beExchangeAsset: {
                beExchangeParentAssetType: 5,
                beExchangeAssetType: helpers_1.CHAIN_INFO.chainAssetType,
                beExchangeAssetPrealnum: "100000",
            },
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createToExchangeAnyMulti(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastToExchangeAnyMulti({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
