// @bfmeta/node-sdk/build/test/transactions/transferAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
(async () => {
    try {
        const secret = "nose install correct solar side latin focus churn mask nominee differ mosquito claw awake glass rare pond clump draw rent fiction muscle razor bacon";
        const keypair = await helpers_1.bfmetaSDK.bfchainSignUtil.createKeypair(secret);
        const publicKey = keypair.publicKey.toString("hex");
        const argv = {
            publicKey,
            fee: "1000",
            applyBlockHeight: 1,
            remark: { message: "create transferAsset" },
            assetType: "BFMTEST",
            amount: "10000",
            recipientId: "cCET2Sxt2LPDhx44wxJ9uhkpviKNrSacvE",
        };
        const sdk = helpers_1.bfmetaSDK;
        const createResult = await sdk.api.transaction.createTransferAsset(argv);
        if (createResult.success) {
            const buffer = createResult.result.buffer;
            const signature = (await helpers_1.bfmetaSDK.bfchainSignUtil.detachedSign(Buffer.from(buffer, "base64"), keypair.secretKey)).toString("hex");
            const broadcastResult = await sdk.api.transaction.broadcastTransferAsset({
                buffer,
                signature,
            });
            console.log(broadcastResult);
        }
        else {
            console.log(createResult);
        }
    }
    catch (e) {
        console.log(e);
    }
})();
