// @bnqkl/util-node/cjs/event/QueneEventEmitterPro.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncQueneEventEmitterPro = exports.QueneEventEmitterPro = void 0;
const tslib_1 = require("tslib");
const index_js_1 = require("../event-base/index.js");
const index_js_2 = require("../event-quene_emitter/index.js");
/**一个极简的事件管理模块 */
class QueneEventEmitterPro extends index_js_2.QueneEventEmitter {
    constructor() {
        super(...arguments);
        //#region on emit any event
        /**是否有过监听通用事件处理 */
        Object.defineProperty(this, "_hasCommonEmitHandlerMap", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
    /**触发事件监听 */
    emit(eventname, data) {
        /**
         * 处理通用监听器
         */
        if (this._hasCommonEmitHandlerMap) {
            if (index_js_1.isDev) {
                if (this._commonEmitHandlerMap.size) {
                    const self = this[index_js_1.EVENT_DESCRIPTION_SYMBOL] || this;
                    console.log(...index_js_1.eventDebugStyle.head("%s QUENE EMIT WILDCARDS * [%s]", index_js_1.eventDebugStyle.DARKVIOLET_BOLD_UNDERLINE), self, eventname, data);
                }
            }
            for (const [commonHanlder, opts] of this._commonEmitHandlerMap) {
                try {
                    if (index_js_1.isDev) {
                        const { taskname = commonHanlder.name } = opts;
                        console.log(...index_js_1.eventDebugStyle.head("RUN [%s]", index_js_1.eventDebugStyle.DARKVIOLET_BOLD_UNDERLINE), taskname);
                    }
                    const res = commonHanlder({ eventname, args: data });
                    if (res instanceof Promise) {
                        res.catch((err) => this._emitErrorHanlder(err, eventname, data));
                    }
                }
                catch (err) {
                    this._emitErrorHanlder(err, eventname, data);
                }
                finally {
                    if (opts.once) {
                        this._commonEmitHandlerMap.delete(commonHanlder);
                    }
                }
            }
        }
        /// 触发事件
        return super.emit(eventname, data);
    }
    get _commonEmitHandlerMap() {
        this._hasCommonEmitHandlerMap = true;
        return new Map();
    }
    /**
     * 监听所有事件
     * @param commonHanlder
     * @param taskname
     */
    onEmit(commonHanlder, opts = {}) {
        if (this._commonEmitHandlerMap.has(commonHanlder)) {
            console.warn(`hanlder '${commonHanlder.name}' already exits in common hanlder event set.`);
        }
        if (opts.taskname === undefined) {
            opts.taskname = (0, index_js_1.GetCallerInfo)(this.onEmit);
        }
        this._commonEmitHandlerMap.set(commonHanlder, opts);
    }
    /**
     * 移除所有事件的监听
     * @param commonHanlder
     */
    offEmit(commonHanlder) {
        if (!this._hasCommonEmitHandlerMap) {
            return false;
        }
        if (commonHanlder) {
            return this._commonEmitHandlerMap.delete(commonHanlder);
        }
        this._commonEmitHandlerMap.clear();
        return true;
    }
    //#endregion
    /**
     * 移除所有的事件
     */
    clear(opts = {}) {
        super.clear(opts);
        const { ignoreCommonErrorHanlder } = opts;
        if (!ignoreCommonErrorHanlder && this._hasCommonEmitHandlerMap) {
            /// 默认清理掉通用监听的回调合集
            this._commonEmitHandlerMap.clear();
        }
    }
}
tslib_1.__decorate([
    index_js_1.cacheGetter,
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [])
], QueneEventEmitterPro.prototype, "_commonEmitHandlerMap", null);
exports.QueneEventEmitterPro = QueneEventEmitterPro;
function syncQueneEventEmitterPro() {
    return new QueneEventEmitterPro();
}
exports.syncQueneEventEmitterPro = syncQueneEventEmitterPro;
