// @bfmeta/node-sdk/build/src/api/atom_api/transactions/migrate_certificate/fromAuthSignatureMigrateCertificateApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FromAuthSignatureMigrateCertificateApi = void 0;
const _migrateCertificateApi_1 = require("./_migrateCertificateApi");
require("../../../../constants");
class FromAuthSignatureMigrateCertificateApi extends _migrateCertificateApi_1.MigrateCertificateApi {
    GENERATE_API_PATH = "/generate/migrateCertificateFromAuthSignature";
    async sendPostRequest(argv) {
        const apiPath = `${this.networkHelper.TRANSACTION_SERVER_URL_PREFIX}${this.GENERATE_API_PATH}`;
        try {
            const result = await this.networkHelper.sendPostRequest(apiPath, argv);
            return result;
        }
        catch (e) {
            const errorInfo = {
                success: false,
                error: {
                    code: "7001",
                    message: `request api ${apiPath} error`,
                    description: e.message,
                },
            };
            return errorInfo;
        }
    }
}
exports.FromAuthSignatureMigrateCertificateApi = FromAuthSignatureMigrateCertificateApi;
