// @bfmeta/node-sdk/build/src/api/atom_api/basics/apis/_basicPostApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicPostApi = void 0;
require("../../../../constants");
class BasicPostApi {
    networkHelper;
    __API_NAMESPACE = "api/basic";
    constructor(networkHelper) {
        this.networkHelper = networkHelper;
    }
    async sendPostRequest(argv) {
        const apiPath = `${this.networkHelper.URL_PREFIX}${this.networkHelper.REQUEST_PROTOCOL === "websocket" ? "post" + "/" : ""}${this.__API_NAMESPACE}${this.REQUEST_API_PATH}`;
        try {
            const result = await this.networkHelper.sendPostRequest(apiPath, argv);
            return result;
        }
        catch (e) {
            const errorInfo = {
                success: false,
                error: {
                    code: "7001",
                    message: `request post api ${apiPath} error`,
                    description: e.message,
                },
            };
            return errorInfo;
        }
    }
}
exports.BasicPostApi = BasicPostApi;
