// @bfmeta/node-sdk/build/src/api/atom_api/basics/apis/_basicGetApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicGetApi = void 0;
require("../../../../constants");
class BasicGetApi {
    networkHelper;
    __API_NAMESPACE = "api/basic";
    constructor(networkHelper) {
        this.networkHelper = networkHelper;
    }
    async sendGetRequest(argv) {
        const apiPath = `${this.networkHelper.URL_PREFIX}${this.networkHelper.REQUEST_PROTOCOL === "websocket" ? "get" + "/" : ""}${this.__API_NAMESPACE}${this.REQUEST_API_PATH}`;
        try {
            const result = await this.networkHelper.sendGetRequest(apiPath, argv);
            return result;
        }
        catch (e) {
            const errorInfo = {
                success: false,
                error: {
                    code: "7001",
                    message: `request get api ${apiPath} error`,
                    description: e.message,
                },
            };
            return errorInfo;
        }
    }
}
exports.BasicGetApi = BasicGetApi;
